src/rules/noVoidExpressionRule.ts

```diff
@@ -36,20 +48,31 @@ export class Rule extends Lint.Rules.TypedRule {
     public static FAILURE_STRING = "Expression has type `void`. Put it on its own line as a statement.";
 
     public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
-        return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
+        const allowArrowFunctionShorthand = this.ruleArguments.indexOf(OPTION_ALLOW_ARROW_FUNCTION_SHORTHAND) !== -1;
+        return this.applyWithFunction(sourceFile, (ctx) => walk(ctx, program.getTypeChecker()), allowArrowFunctionShorthand);
     }
 }
 
-class Walker extends Lint.ProgramAwareRuleWalker {
-    public visitNode(node: ts.Node) {
-        if (isPossiblyVoidExpression(node) && node.parent!.kind !== ts.SyntaxKind.ExpressionStatement && this.isVoid(node)) {
-            this.addFailureAtNode(node, Rule.FAILURE_STRING);
+function walk(ctx: Lint.WalkContext<boolean>, checker: ts.TypeChecker): void {
```

Comment:
a `boolean` walk context looks awkward, let's just make an `IOptions` interface that has one field